README.md
## File Encryptor
__By User0092__

Requirements:
`pip install cryptography`
`pip install pwinput`


changelog
-- V 1.1 --
1. Added `print_text_with_solid_color` to the `color` module to incorporate [Pull #1](https://github.com/User00092/File-Encrypter/pull/1).
2. Updated the `main()` function to include the `handle_response` function.
3. Integrated functions into the actions section of the `main()` function to enhance code organization and minimize potential errors.
4. Renamed `modes` to `actions` to better align with the specific use case.
5. Corrected spacing issues between functions for improved readability.
6. Utilized AI to add comprehensive documentation to functions.
7. Added `try`-`catch` to `main()` within `if __name__ == "__main__"` to handle uncaught exceptions.

color.py
def divide(*numbers: list[float]) -> float | int:
    """
    The `divide` function takes in a variable number of arguments and returns the result of dividing the
    first argument by the rest of the arguments, handling zero division errors by returning 0.0.
    
    :param : The `divide` function takes in a variable number of arguments, which are stored in the
    `numbers` parameter as a list of floats. The function returns either a float or an integer
    :type : list[float]
    :return: The function `divide` returns a float or an integer value.
    """
    try:
        result = numbers[0]
        for number in numbers[1:]:
            result /= number
        return result
    except ZeroDivisionError:
        return 0.0


def create_gradient(start_rgb, end_rgb):
    """
    The function `create_gradient` takes two RGB color values as input and generates a gradient of
    colors between them.
    
    :param start_rgb: The start_rgb parameter is a tuple containing the RGB values for the starting
    color of the gradient
    :param end_rgb: The `end_rgb` parameter represents the ending RGB color of the gradient. It is a
    tuple containing three values: the red component, the green component, and the blue component
    :return: a list of RGB tuples that represent a gradient between the start_rgb and end_rgb colors.
    """
    # Extract the individual RGB components
    start_r, start_g, start_b = start_rgb
    end_r, end_g, end_b = end_rgb

    # Calculate the difference between the start and end RGB values
    diff_r = end_r - start_r
    diff_g = end_g - start_g
    diff_b = end_b - start_b

    # Calculate the number of steps needed for the gradient
    steps = max(abs(diff_r), abs(diff_g), abs(diff_b))

    # Calculate the increment for each RGB component
    increment_r = divide(diff_r, steps)
    increment_g = divide(diff_g, steps)
    increment_b = divide(diff_b, steps)

    # Generate the gradient by incrementing the RGB values
    gradient = []
    for i in range(steps + 1):
        r = int(start_r + i * increment_r)
        g = int(start_g + i * increment_g)
        b = int(start_b + i * increment_b)
        gradient.append((r, g, b))

    return gradient

def get_text_with_gradient(text, gradient):
    """
    The function `get_text_with_gradient` takes a text and a gradient as input and returns the text with
    each character formatted with a corresponding color from the gradient.
    
    :param text: The text parameter is a string that represents the text that you want to format with a
    gradient
    :param gradient: The gradient parameter is a list of RGB values. Each RGB value represents a color
    in the gradient. The length of the gradient list determines the number of colors in the gradient
    :return: the formatted text with a gradient effect.
    """
    text_length = len(text)
    gradient_length = len(gradient)

    # Calculate the ratio between the text length and gradient length
    ratio = divide(gradient_length, text_length)

    # Initialize an empty string to store the formatted text
    formatted_text = ""

    # Iterate over each character in the text
    for i, char in enumerate(text):
        gradient_index = int(i * ratio)
        rgb = gradient[gradient_index]
        rgb_string = f"{rgb[0]};{rgb[1]};{rgb[2]}"
        formatted_char = f"\033[38;2;{rgb_string}m{char}\033[0m"
        formatted_text += formatted_char

    return formatted_text

def print_text_with_solid_color(text: str, color: tuple | list):
    """
    The function `print_text_with_solid_color` takes a string and a color as input and prints the string
    with the specified color.
    
    :param text: The `text` parameter is a string that represents the text you want to print with a
    solid color
    :type text: str
    :param color: The `color` parameter is a tuple or list that represents the RGB values of the desired
    color. It should contain three integers ranging from 0 to 255, representing the red, green, and blue
    components of the color respectively
    :type color: tuple | list
    """
    rgb_string = f"{color[0]};{color[1]};{color[2]}"
    formatted_text = f"\033[38;2;{rgb_string}m{text}\033[0m"
    print(formatted_text)


utils.py
from tkinter.filedialog import askopenfilename, askdirectory
from cryptography.fernet import Fernet
import platform
import pwinput
import tkinter
import base64
import json
import os

encrypted_file_extension = ".user_enc"
sys_os = platform.system()


def clear_terminal() -> None:
    """
    The function clears the terminal screen.
    """
    os.system("cls") if sys_os == "Windows" else os.system('clear')


def can_file_be_renamed(path: str) -> bool:
    """
    The function `can_file_be_renamed` checks if a file can be renamed by attempting to rename it and
    returning True if successful, False otherwise.
    
    :param path: The `path` parameter is a string that represents the file path of the file that you
    want to check if it can be renamed
    :type path: str
    :return: a boolean value. It returns True if the file at the given path can be renamed, and False if
    it cannot be renamed.
    """
    try:
        os.rename(path, path)
        return True
    except Exception:
        return False


def can_file_be_written(path: str) -> bool:
    """
    The function `can_file_be_written` checks if a file can be opened for reading and writing.
    
    :param path: The `path` parameter is a string that represents the file path of the file you want to
    check if it can be written
    :type path: str
    :return: The function `can_file_be_written` returns a boolean value. It returns `True` if the file
    at the given path can be successfully opened and written to, and `False` if there is an exception or
    error encountered during the process.
    """
    try:
        with open(path, "rb") as f:
            data = f.read()
        
        with open(path, 'wb') as f:
            f.write(data)
        
        return True
    except Exception:
        return False

def encrypt(plaintext: str | bytes, key: str | bytes, return_bytes: bool = False) -> str | bytes:
    """
    The `encrypt` function takes a plaintext message and a key, and encrypts the message using the
    Fernet encryption algorithm, returning the encrypted message as a string or bytes depending on the
    `return_bytes` parameter.
    
    :param plaintext: The `plaintext` parameter represents the text or data that you want to encrypt. It
    can be either a string or bytes
    :type plaintext: str | bytes
    :param key: The `key` parameter is the encryption key used to encrypt the plaintext. It can be
    either a string or bytes. If it is a string, it will be encoded into bytes before encryption
    :type key: str | bytes
    :param return_bytes: The `return_bytes` parameter is a boolean flag that determines whether the
    encrypted result should be returned as bytes or as a string. If `return_bytes` is set to `True`, the
    function will return the encrypted result as bytes. If `return_bytes` is set to `False` (or,
    defaults to False
    :type return_bytes: bool (optional)
    :return: The function `encrypt` returns either a string or bytes, depending on the value of the
    `return_bytes` parameter. If `return_bytes` is `True`, it returns the encrypted result as bytes.
    Otherwise, it returns the encrypted result as a string.
    """
    try:
        if isinstance(key, str):
            key = key.encode()

        if not isinstance(plaintext, str) and not isinstance(plaintext, bytes):
            plaintext = str(plaintext)

        if isinstance(plaintext, str):
            plaintext = plaintext.encode()

        fernet = Fernet(key)
        result = fernet.encrypt(plaintext)

        if return_bytes:
            return result
        
        return result.decode()
    except Exception:
        return None


def decrypt(ciphertext: str | bytes, key: str | bytes, return_bytes: bool = False) -> str | bytes | None:
    """
    The `decrypt` function takes a ciphertext and a key, and attempts to decrypt the ciphertext using
    the key, returning the decrypted result as a string or bytes object, or None if decryption fails.
    
    :param ciphertext: The `ciphertext` parameter represents the encrypted message that needs to be
    decrypted. It can be either a string or bytes
    :type ciphertext: str | bytes
    :param key: The `key` parameter is the encryption key used to decrypt the ciphertext. It can be
    either a string or bytes. If it is a string, it will be encoded to bytes before decryption
    :type key: str | bytes
    :param return_bytes: The `return_bytes` parameter is a boolean flag that determines whether the
    decrypted result should be returned as bytes or as a string. If `return_bytes` is set to `True`, the
    function will return the decrypted result as bytes. If `return_bytes` is set to `False` (or,
    defaults to False
    :type return_bytes: bool (optional)
    :return: The function `decrypt` returns either a string, bytes, or `None`.
    """
    try:
        if isinstance(key, str):
            key = key.encode()
        
        if isinstance(ciphertext, str):
            ciphertext = base64.b64decode(ciphertext)
        
        try:
            fernet = Fernet(key)
            result = fernet.decrypt(ciphertext)
        except Exception:
            return None

        if return_bytes:
            return result
        
        return result.decode()
    except Exception:
        return None


def get_hidden_input(prompt: str, mask: str = "*") -> str:
    """
    The function `get_hidden_input` takes a prompt and an optional mask character as input, and returns
    a hidden input from the user.
    
    :param prompt: The prompt is a string that will be displayed to the user to ask for input. It is
    used to provide instructions or ask a question to the user
    :type prompt: str
    :param mask: The `mask` parameter is a string that specifies the character(s) to be used as a mask
    for hiding the input. By default, it is set to "*", defaults to *
    :type mask: str (optional)
    :return: a string.
    """
    return pwinput.pwinput(prompt=prompt, mask=mask)


# The `Config` class is a Python class that provides methods for reading and writing configuration
# data from a JSON file.
class Config:
    def __init__(self, config_path: str) -> None:
        self.config_path = config_path
        self._config = self.load_config()

    def _fix_config(self) -> None:
        """
        The function `_fix_config` checks if a config file exists, and if not, creates an empty one.
        :return: None.
        """
        if not os.path.exists(self.config_path):
            with open(self.config_path, 'w') as f:
                json.dump({}, f)
            
            return
            
        with open(self.config_path, 'r') as f:
            if f.read().strip() == '':
                with open(self.config_path, 'w') as f:
                    json.dump({}, f)
        
    def load_config(self):
        """
        The function loads a JSON configuration file and assigns it to the `_config` attribute after
        fixing it.
        """
        self._fix_config()
        with open(self.config_path, 'r') as f:
            self._config = json.load(f)
        
    def read(self) -> dict:
        """
        The function reads and returns the loaded configuration as a dictionary.
        :return: The method is returning a dictionary object.
        """
        self.load_config()
        return self._config
    
    def write(self, config: dict) -> None:
        """
        The function writes a dictionary object to a file in JSON format.
        
        :param config: The `config` parameter is a dictionary that contains the configuration data that
        you want to write to a file
        :type config: dict
        """
        self._config = config
        with open(self.config_path, 'w') as f:
            json.dump(self._config, f)


def list_files(directory: str) -> set:
    """
    The function `list_files` recursively lists all files in a given directory and its subdirectories.
    
    :param directory: The directory parameter is a string that represents the path to a directory
    :type directory: str
    :return: a set of file paths within the specified directory and its subdirectories.
    """
    files = set()
    
    for file in os.listdir(directory):
        path = os.path.join(directory, file)
        if os.path.isdir(path):
            for file in list_files(path):
                files.add(file)
        else:
            files.add(path)
    
    return files


def encrypt_file(file_path: str, key: str) -> bool:
    """
    The function encrypts a file using a given key and renames the file with an encrypted file
    extension.
    
    :param file_path: The `file_path` parameter is a string that represents the path to the file that
    needs to be encrypted
    :type file_path: str
    :param key: The `key` parameter is a string that represents the encryption key. This key is used to
    encrypt the data in the file
    :type key: str
    :return: a boolean value. It returns True if the file was successfully encrypted and renamed, and
    False otherwise.
    """
    if file_path.endswith(encrypted_file_extension):
        return False
    
    if not can_file_be_renamed(file_path) or not can_file_be_written(file_path):
        return False
    
    with open(file_path, 'rb') as f:
        data = f.read()
    
    encrypted_data = encrypt(data, key, True)
    
    with open(file_path, 'wb') as f:
        f.write(encrypted_data)

    os.rename(file_path, file_path + encrypted_file_extension)
    return True


def decrypt_file(file_path: str, key: str) -> bool:
    """
    The function decrypts a file using a given key and renames it to remove the encrypted file
    extension.
    
    :param file_path: The `file_path` parameter is a string that represents the path to the file that
    needs to be decrypted
    :type file_path: str
    :param key: The "key" parameter is a string that represents the encryption key used to decrypt the
    file
    :type key: str
    :return: a boolean value. It returns True if the file was successfully decrypted and renamed, and
    False otherwise.
    """
    if not file_path.endswith(encrypted_file_extension):
        return False
    
    if not can_file_be_renamed(file_path) or not can_file_be_written(file_path):
        return False
    
    with open(file_path, 'rb') as f:
        data = f.read()
    
    decrypted_data = decrypt(data, key, True)
    if decrypted_data is None:
        return False
    
    with open(file_path, 'wb') as f:
        f.write(decrypted_data)

    os.rename(file_path, file_path.removesuffix(encrypted_file_extension))
    return True


def select_file() -> str | None:
    """
    The `select_file` function opens a file dialog box for the user to select a file and returns the
    path of the selected file or `None` if no file is selected.
    :return: The function `select_file()` returns a string representing the path of the selected file,
    or `None` if no file was selected.
    """
    root = tkinter.Tk()
    root.title("Select File")
    root.geometry('0x0')
    path = askopenfilename() or None
    root.destroy()
    return path

def select_directory() -> str | None:
    """
    The function `select_directory` opens a file dialog to allow the user to select a directory and
    returns the selected directory path or `None` if no directory is selected.
    :return: The function `select_directory` returns a string representing the selected directory path,
    or `None` if no directory was selected.
    """
    root = tkinter.Tk()
    root.title("Select Directory")
    root.geometry('0x0')
    path = askdirectory() or None
    root.destroy()
    return path


actions.py
import utils
import color
import sys
import os

config = utils.Config("config.json")
gradient = color.create_gradient((65, 105, 225), (65, 105, 225))
info_gradient_color = 191, 0, 255

directory_gradient = color.create_gradient((0, 128, 0), (0, 128, 0))
file_gradient = color.create_gradient((0, 0, 255), (0, 0, 255))


def get_cryption_key() -> bytes | None:
    """
    The function `get_cryption_key` takes user input for a key, checks its length, encodes it using
    base64, and returns the encoded key if it is valid, otherwise it returns None.
    :return: The function `get_cryption_key()` returns either a `bytes` object or `None`.
    """
    try:
        key_input: str = utils.get_hidden_input("Enter your key: ")
        key_len: int = len(key_input)
        if key_len < 4 or key_len > 32:
            return {
                "status": False,
                "message": "Invalid key length! Must be no more than 32 characters."
            }

        key_input = utils.base64.urlsafe_b64encode((key_input + ("w" * (32 - key_len))).encode('utf-8'))

        utils.Fernet(key_input)
    
        return key_input
    except Exception:
        return None


def encrypt_paths(paths: tuple) -> dict:
    """
    The `encrypt_paths` function takes a tuple of file paths, encrypts each file using a provided
    encryption key, and returns a dictionary with the status and message of the encryption process.
    
    :param paths: A tuple containing the paths of files or directories to be encrypted
    :type paths: tuple
    :return: The function `encrypt_paths` returns a dictionary with two keys: "status" and "message".
    The "status" key indicates whether the encryption process was successful or not, and the "message"
    key provides a description of the result or any error messages.
    """
    try:
        if isinstance(paths, str):
            return {
                "status": False,
                "message": "Config is corrupt."
            }
        
        key_input: dict = get_cryption_key()
        if not key_input:
            return {
                "status": False,
                "message": "Key cannot be used."
            }
        
        color.print_text_with_solid_color("Please be patient... Process may seem frozen.", info_gradient_color)

        file_paths = list()
        for path in paths:
            filep = os.path.abspath(path)
            if os.path.isdir(path):
                for file in utils.list_files(filep):
                    filep = os.path.abspath(file)
                    file_paths.append(filep)
            else:
                file_paths.append(filep)
        
        file_path_count = len(file_paths)
        processed_paths = 0

        for file in file_paths:
            result = utils.encrypt_file(file, key_input)
            if result:
                processed_paths += 1

        
            sys.stdout.write(color.get_text_with_gradient(f'Encrypting files... {round(color.divide(processed_paths, file_path_count) * 100, 2)}% ({processed_paths} of {file_path_count})', gradient))
            sys.stdout.write('\r')
            sys.stdout.flush()
        print()
        return {
            "status": True,
            "message": "Successfully encrypted files."
        }
    except Exception:
        print()
        return {
            "status": False,
            "message": "An unexpected error occurred."
        }


def decrypt_paths(paths: tuple) -> dict:
    """
    The `decrypt_paths` function takes a tuple of file paths, decrypts each file using a provided key,
    and returns a dictionary with the status and message of the decryption process.
    
    :param paths: A tuple containing file paths to be decrypted
    :type paths: tuple
    :return: The function `decrypt_paths` returns a dictionary with two keys: "status" and "message".
    The "status" key indicates whether the decryption process was successful or not, and the "message"
    key provides a description of the result or any error messages.
    """
    try:
        if isinstance(paths, str):
            return {
                "status": False,
                "message": "Config is corrupt."
            }
        
        key_input: dict = get_cryption_key()
        if not key_input:
            return {
                "status": False,
                "message": "Key cannot be used."
            }
        
        color.print_text_with_solid_color("Please be patient... Process may seem frozen.", info_gradient_color)

        file_paths = list()
        for path in paths:
            filep = os.path.abspath(path)
            if os.path.isdir(path):
                for file in utils.list_files(filep):
                    filep = os.path.abspath(file)
                    file_paths.append(filep)
            else:
                file_paths.append(filep)
        
        file_path_count = len(file_paths)
        processed_paths = 0

        for file in file_paths:
            result = utils.decrypt_file(file, key_input)
            if result:
                processed_paths += 1

            sys.stdout.write(color.get_text_with_gradient(f'Decrypting files... {round(color.divide(processed_paths, file_path_count) * 100, 2)}% ({processed_paths} of {file_path_count})', gradient))
            sys.stdout.write('\r')
            sys.stdout.flush()
        
        print()
        return {
            "status": True,
            "message": "Successfully decrypted files."
        }
    except Exception:
        print()
        return {
            "status": False,
            "message": "An unexpected error occurred."
        }


def add_file_to_config() -> dict:
    """
    The function `add_file_to_config()` adds a file path to a configuration file if it meets certain
    conditions.
    :return: The function `add_file_to_config()` returns a dictionary with two keys: "status" and
    "message". The "status" key indicates whether the operation was successful or not, with a value of
    `True` for success and `False` for failure. The "message" key provides a descriptive message about
    the result of the operation.
    """
    color.print_text_with_solid_color("Please check your task bar for the prompt.", info_gradient_color)
    file_path = utils.select_file()
    if not file_path:
        return {
            "status": False,
            "message": "No file selected."
        }
    
    if file_path.endswith(utils.encrypted_file_extension):
        return {
            "status": False,
            "message": "This file is already encrypted."
        }
    
    config_data: dict = config.read()

    if config_data.get("paths") is None:
        config_data["paths"] = list()
    
    if file_path in config_data["paths"]:
        return {
            "status": False,
            "message": "File already exists in config."
        }
    
    config_data["paths"].append(file_path)
    config.write(config_data)

    return {
        "status": True,
        "message": "Successfully added file to config."
    }


def remove_file_from_config() -> dict:
    """
    The function `remove_file_from_config` removes a file path from a configuration file.
    :return: The function `remove_file_from_config()` returns a dictionary with two keys: "status" and
    "message". The "status" key indicates whether the file removal was successful (True) or not (False).
    The "message" key provides a corresponding message explaining the result of the file removal
    operation.
    """
    color.print_text_with_solid_color("Please check your task bar for the prompt.", info_gradient_color)
    file_path = utils.select_file()
    if not file_path:
        return {
            "status": False,
            "message": "No file selected."
        }

    config_data: dict = config.read()

    if config_data.get("paths") is None:
        config_data["paths"] = list()
    
    if file_path not in config_data["paths"]:
        return {
            "status": False,
            "message": "File does not exists in config."
        }
    
    config_data["paths"].remove(file_path)
    config.write(config_data)

    return {
        "status": True,
        "message": "Successfully removed file from config."
    }


def add_folder_to_config() -> dict:
    """
    The function `add_folder_to_config` adds a selected directory path to a configuration file.
    :return: The function `add_folder_to_config()` returns a dictionary with two keys: "status" and
    "message". The "status" key indicates whether the operation was successful or not, with a value of
    `True` for success and `False` for failure. The "message" key provides a descriptive message about
    the outcome of the operation.
    """
    color.print_text_with_solid_color("Please check your task bar for the prompt.", info_gradient_color)
    directory_path = utils.select_directory()
    if not directory_path:
        return {
            "status": False,
            "message": "No directory selected."
        }
    
    config_data: dict = config.read()

    if config_data.get("paths") is None:
        config_data["paths"] = list()
    
    if directory_path in config_data["paths"]:
        return {
            "status": False,
            "message": "Directory already exists in config."
        }
    
    config_data["paths"].append(directory_path)
    config.write(config_data)

    return {
        "status": True,
        "message": "Successfully added directory to config."
    }


def remove_folder_from_config() -> dict:
    """
    The function `remove_folder_from_config` removes a directory path from a configuration file.
    :return: The function `remove_folder_from_config()` returns a dictionary with two keys: "status" and
    "message". The "status" key indicates whether the operation was successful or not, with a value of
    `True` for success and `False` for failure. The "message" key provides a descriptive message about
    the outcome of the operation.
    """
    color.print_text_with_solid_color("Please check your task bar for the prompt.", info_gradient_color)
    directory_path = utils.select_directory()
    if not directory_path:
        return {
            "status": False,
            "message": "No directory selected."
        }

    config_data: dict = config.read()

    if config_data.get("paths") is None:
        config_data["paths"] = list()
    
    if directory_path not in config_data["paths"]:
        return {
            "status": False,
            "message": "Directory does not exists in config."
        }
    
    config_data["paths"].remove(directory_path)
    config.write(config_data)

    return {
        "status": True,
        "message": "Successfully removed directory from config."
    }


def list_paths_in_config() -> dict:
    """
    The function `list_paths_in_config` reads the config data, checks if the "paths" key exists, adds it
    if it doesn't, and then prints the paths in the config with different colors depending on whether
    they are directories or files.
    """
    config_data: dict = config.read()

    if config_data.get("paths") is None:
        config_data["paths"] = list()
        config.write(config_data)
    
    color.print_text_with_solid_color("Paths in config:", info_gradient_color)
    for path in config_data["paths"]:
        if not path:
            continue

        if os.path.isdir(path):
            print("\t" + color.get_text_with_gradient(path, directory_gradient))
        else:
            print("\t" + color.get_text_with_gradient(path, file_gradient))


main.py
import actions
import utils
import color

config = utils.Config("config.json")
green_rgb = (128, 202, 49)
red_rgb = (224, 17, 43)
royal_blue_rgb = (65, 105, 225)


def main() -> bool:
    def handle_response(response: dict) -> bool:
        color.print_text_with_solid_color(response.get('message', "No message provided."), green_rgb if response.get('status') else red_rgb)
    
    action_choices: dict = {
        1: {"name": "Encrypt", "function": lambda: handle_response(actions.encrypt_paths(config.read().get("paths", tuple())))},
        2: {"name": "Decrypt", "function": lambda: handle_response(actions.decrypt_paths(config.read().get("paths", tuple())))},
        3: {"name": "Add file", "function": lambda: handle_response(actions.add_file_to_config())},
        4: {"name": "Remove file", "function": lambda: handle_response(actions.remove_file_from_config())},
        5: {"name": "Add folder", "function": lambda: handle_response(actions.add_folder_to_config())},
        6: {"name": "Remove folder", "function": lambda: handle_response(actions.remove_folder_from_config())},
        7: {"name": "List paths", "function": lambda: actions.list_paths_in_config()},
        8: {"name": "Exit"}
    }

    display_text = "Select an action: "

    for mode, value in action_choices.items():
        display_text += f"\n{mode}: {value.get('name')}"

    color.print_text_with_solid_color(display_text, royal_blue_rgb)
    
    selected_mode: str = input(color.get_text_with_gradient("--> ", (royal_blue_rgb, royal_blue_rgb)))

    if not selected_mode or not selected_mode.isdigit() or int(selected_mode) not in action_choices:
        print("Please enter a valid choice.")
        return True
    
    selected_mode = int(selected_mode)
    
    function_ = action_choices[selected_mode].get("function")
    if function_ is None:
        return False
    else:
        function_()
        return True


if __name__ == "__main__":
    while True:
        try:
            utils.clear_terminal()
            try:
                result = main()
                if not result:
                    break
            except Exception:
                utils.clear_terminal()
                color.print_text_with_solid_color("Something went wrong.", red_rgb)
            
            input("Press enter to continue...")
        except KeyboardInterrupt:
            break
    
    utils.clear_terminal()
    color.print_text_with_solid_color("Exited.", red_rgb)
